README.md

```diff
@@ -211,42 +212,163 @@ val records = sc.newAPIHadoopFile(
+    Spark connects to S3 using both the Hadoop FileSystem interfaces and directly using the Amazon
+    Java SDK's S3 client.
+
+    This connection can be authenticated using either AWS keys or IAM roles (DBFS mountpoints are
+    not currently supported, so Databricks users who do not want to rely on AWS keys should use
+    cluster IAM roles instead).
+
+    There are multiple ways of providing these credentials:
+
+    1. **Default Credential Provider Chain (best option for most users):**
+        AWS credentials will automatically be retrieved through the [DefaultAWSCredentialsProviderChain](http://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html#id6).
+
+        If you use IAM instance roles to authenticate to S3 (e.g. on Databricks, EMR, or EC2), then
+        you should probably use this method.
+
+        If another method of providing credentials is used (methods 2 or 3), then that will take
+        precedence over this default.
+
+    2. **Set keys in Hadoop conf:** You can specify AWS keys via
+        [Hadoop configuration properties](https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/index.md).
+        For example, if your `tempdir` configuration points to a `s3n://` filesystem then you can
+        set the `fs.s3n.awsAccessKeyId` and `fs.s3n.awsSecretAccessKey` properties in a Hadoop XML
+        configuration file or call `sc.hadoopConfiguration.set()` to mutate Spark's global Hadoop
+        configuration.
+
+        For example, if you are using the `s3n` filesystem then add
+
+        ```scala
+        sc.hadoopConfiguration.set("fs.s3n.awsAccessKeyId", "YOUR_KEY_ID")
+        sc.hadoopConfiguration.set("fs.s3n.awsSecretAccessKey", "YOUR_SECRET_ACCESS_KEY")
+        ```
+
+        and for the `s3a` filesystem add
+
+        ```scala
+        sc.hadoopConfiguration.set("fs.s3a.access.key", "YOUR_KEY_ID")
+        sc.hadoopConfiguration.set("fs.s3a.secret.key", "YOUR_SECRET_ACCESS_KEY")
+        ```
+
+        Python users will have to use a slightly different method to modify the `hadoopConfiguration`,
+        since this field is not exposed in all versions of PySpark. Although the following command
+        relies on some Spark internals, it should work with all PySpark versions and is unlikely to
+        break or change in the future:
+
+        ```python
+        sc._jsc.hadoopConfiguration().set("fs.s3n.awsAccessKeyId", "YOUR_KEY_ID")
+        sc._jsc.hadoopConfiguration().set("fs.s3n.awsSecretAccessKey", "YOUR_SECRET_ACCESS_KEY")
+        ```
+
+    3. **Encode keys in `tempdir` URI**:
+     For example, the URI `s3n://ACCESSKEY:SECRETKEY@bucket/path/to/temp/dir` encodes the key pair
+      (`ACCESSKEY`, `SECRETKEY`).
+
+      Due to [Hadoop limitations](https://issues.apache.org/jira/browse/HADOOP-3733), this
+      approach will not work for secret keys which contain forward slash (`/`) characters.
```

Comment:
I assume most users are familiar with the need to urlencode s3 urls as the workaround to this type of issue; however since that doesn't work maybe mention that urlencode won't fix this issue (for people like me who didn't open the link on the first read through :D)